### ISSUE TITLE
Tracking down PMIX_ERR_OUT_OF_RESOURCE problem

### ISSUE THREAD

**Issue by krempel-pt** [openpmix/openpmix#3433]
### What version of the PMIx Reference Library are you using?
5.0.3

### Describe how PMIx was installed
release tarball

  Configure command line: '--build=x86_64-redhat-linux-gnu'                          
                          '--host=x86_64-redhat-linux-gnu'                           
                          '--program-prefix=' '--disable-dependency-tracking'        
                          '--prefix=/usr' '--exec-prefix=/usr'                       
                          '--bindir=/usr/bin' '--sbindir=/usr/sbin'         
                          '--sysconfdir=/etc' '--datadir=/usr/share'                 
                          '--includedir=/usr/include' '--libdir=/usr/lib64'          
                          '--libexecdir=/usr/libexec' '--localstatedir=/var'         
                          '--sharedstatedir=/var/lib'                                
                          '--mandir=/usr/share/man'                       
                          '--infodir=/usr/share/info' '--prefix=/usr'      
                          '--sysconfdir=/etc/pmix' '--disable-static'       
                          '--disable-silent-rules' '--enable-shared'       
                          '--with-munge'                               

### Please describe the system on which you are running

* Operating system/version: Rocky Linux 8
* Computer hardware: VM qemu/KVM x86_64
* Network type: virtio

-----------------------------

## Details of the problem

I did once again setup a new virtual cluster and stumbled over an old friend.

Actually, it is exactly the same as described in https://github.com/openpmix/openpmix/issues/3204#issuecomment-1835926620

If I just install everything and try to use PMIx, PMIx_Init returns with:
```
[compute-00:29946] PMIX ERROR: PMIX_ERR_OUT_OF_RESOURCE in file client/pmix_client.c at line 278
[compute-00:29948] PMIX ERROR: PMIX_ERR_OUT_OF_RESOURCE in file client/pmix_client.c at line 278
[pspmix_0x0001000074e7[1:29927]:0]: PMIx_Init failed: PMIX_ERROR
[pspmix_0x0001000074e7[1:29927]:1]: PMIx_Init failed: PMIX_ERROR
srun: error: compute-00: tasks 0-1: Exited with exit code 1
```

Setting `PMIX_MCA_gds=hash` for the client only, the error disappears and everything is working.

I would like to try to track this down again, to get rid of the necessity to remember to set this on each new installation. I wonder if this could be somehow related to the virtual environment, but have no ideas yet.

Any help, hints or explanations would be appreciated.

Many thanks in advance.

**Comment by rhc54** [comment]
Not sure why constraining the client to "hash" makes a difference - could be that the VM and shmem component are conflicting so badly that memory is having trouble. The error stems from here:
```c
            sdest[i] = (char *) malloc(len); // NULL terminator is included
            if (NULL == sdest[i]) {
                return PMIX_ERR_OUT_OF_RESOURCE;
            }
```
So a basic `malloc` operation is failing. I'm not sure there is anything we can do about it - you are hitting this when we send you job data (which in this case is the shmem connect info, and you are unpacking the nspace identifying whose job info is being referenced). This occurs long after the GDS has been setup, so there is no way we can "undo" anything.

What I usually do for such a situation is create a platform file that contains (among other things) a default MCA param file with the necessary/desirable settings for that environment. Unfortunately, that only helps when you build the project - doesn't help with an rpm install.

Afraid that leaves me with no great ideas for your situation. 😞

 I suppose one thing you could check (just to ensure something isn't borked) is that "len" value - is it a legitimate value, or did we unpack trash for the length of the string? The code is in src/mca/bfrops/base/bfrop_base_unpack.c, starting at line 402 (on the head of the PMIx v5.0 branch).

**Comment by krempel-pt** [comment]
Thanks Ralph for the insight, I am still interested in digging into that, just did not find the time to do so, yet.

**Comment by rhc54** [comment]
Only thing I can advise is to reiterate the suggestion to add a print statement outputting the value of "len" in the above code. Could be that it is a bad value, perhaps from some code path that fails to initialize it. If it is valid, then I'm not quite sure what we do about it - would require some thought.

**Comment by krempel-pt** [comment]
It looks like the garbage guess was right, printing `len` above the `malloc` reveals:
```
[node01:13698] PMIX ERROR: PMIX_ERR_OUT_OF_RESOURCE in file client/pmix_client.c at line 278
[node01:13696] PMIX ERROR: PMIX_ERR_OUT_OF_RESOURCE in file client/pmix_client.c at line 278
****** len = 12 *******
****** len = 14 *******
****** len = 13 *******
****** len = 25 *******
****** len = 15 *******
****** len = 31 *******
****** len = 31 *******
****** len = 31 *******
****** len = -27 *******
[pspmix_0x00000000357a[0:13690]:0]: PMIx_Init failed: PMIX_ERROR
****** len = 12 *******
****** len = 14 *******
****** len = 13 *******
****** len = 25 *******
****** len = 15 *******
****** len = 31 *******
****** len = 31 *******
****** len = 31 *******
****** len = -27 *******
[pspmix_0x00000000357a[0:13690]:2]: PMIx_Init failed: PMIX_ERROR
PSIlogger: Child with rank 0 exited with status 1.
PSIlogger: Child with rank 2 exited with status 1.
```

**Comment by rhc54** [comment]
Interesting - thx! I'll try to see if I can spot the problem. If necessary, I'll come back to ask for some more debug output.

**Comment by rhc54** [comment]
Afraid I need a little more help here. Could you give me some idea of the setup that generated that output? I'm particularly looking for how many procs were executing? Were they all on the same node?

I see that ranks 0 and 2 exited with error - were the other "len" outputs from different procs? Guess what I'm trying to figure out is if we fail every time we flow through that specific code path (starting in the client when it receives the job info), or are we only sometimes hitting that failure (i.e., some procs succeed while others fail).

Nothing in the code path immediately jumps out at me. One possibility we sometimes see is if the server is using debug buffers (which includes extra data in them) while the client is using non-debug (or "optimized") buffers that does not include the extra data. If that happens, then the client starts unpacking garbage as it has no way to detect that they have a mismatch in communication. But then every proc would behave the same way.

**Comment by rhc54** [comment]
I'm not seeing any uninitialized variables in my code review. If every proc is behaving the same, and we don't have a debug vs optimized mismatch, then I suspect the issue must be a string that isn't NULL terminated (which would cause `strlen` to go bonkers).

Perhaps another approach here would be (if possible) for you to provide me with instructions on how to build/run this VM your using to reproduce this behavior? Would be easier for us to debug it - the next step is likely to be adding a bunch of debug output to see what's going on.

**Comment by rhc54** [comment]
Might be - the key is to look at the corresponding "pack" to see what length it thought the string was when packing. Patch like the following should show it:
```diff
diff --git a/src/mca/bfrops/base/bfrop_base_pack.c b/src/mca/bfrops/base/bfrop_base_pack.c
index 7997ada7..0914a14d 100644
--- a/src/mca/bfrops/base/bfrop_base_pack.c
+++ b/src/mca/bfrops/base/bfrop_base_pack.c
@@ -270,6 +270,7 @@ pmix_status_t pmix_bfrops_base_pack_string(pmix_pointer_array_t *regtypes, pmix_
             }
         } else {
             len = (int32_t) strlen(ssrc[i]) + 1; // retain the NULL terminator
+            pmix_output(0, "PACK STRING %s LEN %d", ssrc[i], len);
             PMIX_BFROPS_PACK_TYPE(ret, buffer, &len, 1, PMIX_INT32, regtypes);
             if (PMIX_SUCCESS != ret) {
                 return ret;
```

**Comment by krempel-pt** [comment]
I did your suggested output on the pack side, and I guess this points not to an invalid passed string, right?

server:
```
PACK STRING pspmix_1001 LEN 12                                                                                                                                            
PACK STRING pmix.evnondef LEN 14                                                                                                                                          
PACK STRING pmix.pset.nm LEN 13                                                                                                                                           
PACK STRING pspmix:reservation/17455 LEN 25                                                                                                                               
PACK STRING pmix.pset.mems LEN 15                                                                                                                                         
PACK STRING pspmix_0x0000000043f0[0:17392] LEN 31                                                                                                                         
PACK STRING pmix.srv.nspace LEN 16                                                                                                                                        
PACK STRING pspmix_1001 LEN 12                                                                                                                                            
PACK STRING pmix.srv.rank LEN 14                                                                                                                                          
PACK STRING pmix.job.size LEN 14                                                                                                                                          
PACK STRING pmix.ssn.arr LEN 13                                                                                                                                           
PACK STRING pmix.clid LEN 10                                                                                                                                              
PACK STRING ParaStationCluster LEN 19                                                                                                                                     
PACK STRING pmix.session.id LEN 16
```
client:
```
[krempel@node01 ~]$ mpiexec --pmix -N 0 -P -F -np 1 TestPrograms/PMIx_Init
****** len = 12 *******
****** sdest[i] = pspmix_1001 *******
****** len = 14 *******
****** sdest[i] = pmix.evnondef *******
****** len = 13 *******
****** sdest[i] = pmix.pset.nm *******
****** len = 25 *******
****** sdest[i] = pspmix:reservation/17455 *******
****** len = 15 *******
****** sdest[i] = pmix.pset.mems *******
****** len = 31 *******
****** sdest[i] = pspmix_0x0000000043f0[0:17392] *******
****** len = -27 *******
[node01:17396] PMIX ERROR: PMIX_ERR_OUT_OF_RESOURCE in file client/pmix_client.c at line 278
[pspmix_0x0000000043f0[0:17392]:0]: PMIx_Init failed: PMIX_ERROR
PSIlogger: Child with rank 0 exited with status 1.
```

**Comment by rhc54** [comment]
Looks like the inputs are okay. Let me just recap a bit to verify we are on the same page.

You see the problem when the server is using PMIx v5.0.6 while the client is using v5.0.3. However, you also see it when both are using v5.0.3 (per your initial description here). Correct? Is this also true if both use v5.0.6?

And to be safe: when client and server are using different versions, those versions were built with the identical configure line?

Just trying to poke at the beast as I have absolutely no idea what might be causing the problem.

**Comment by rhc54** [comment]
Couple of other quick thoughts:

1. Add `--enable-debug` to your PMIx configuration(s). Let's see if we are hitting a pack type mismatch
2. Can you please walk me through your VM setup? I'd like to try and locally reproduce this so I can more efficiently track it down

**Comment by krempel-pt** [comment]
> Looks like the inputs are okay. Let me just recap a bit to verify we are on the same page.
> 
> You see the problem when the server is using PMIx v5.0.6 while the client is using v5.0.3. However, you also see it when both are using v5.0.3 (per your initial description here). Correct? Is this also true if both use v5.0.6?

Yes, yes, and yes. In fact the outputs above are all generated with 5.0.6 only and both server and client using the same build.

> 
> And to be safe: when client and server are using different versions, those versions were built with the identical configure line?

No that was not necessarily the case. But sometimes, yes.


> 
> Just trying to poke at the beast as I have absolutely no idea what might be causing the problem.

Go ahead.

**Comment by krempel-pt** [comment]
> Couple of other quick thoughts:
> 
> 1. Add `--enable-debug` to your PMIx configuration(s). Let's see if we are hitting a pack type mismatch

Yes, we do. Actually, I already tried that and wanted to add that as well, but then forgot it at the end:

```
krempel@node01 ~]$ mpiexec --pmix -N 0 -P -F -np 1 TestPrograms/PMIx_Init
****** len = 12 *******
****** sdest[i] = pspmix_1001 *******
****** len = 14 *******
****** sdest[i] = pmix.evnondef *******
****** len = 13 *******
****** sdest[i] = pmix.pset.nm *******
****** len = 25 *******
****** sdest[i] = pspmix:reservation/17694 *******
****** len = 15 *******
****** sdest[i] = pmix.pset.mems *******
****** len = 31 *******
****** sdest[i] = pspmix_0x0000000044df[0:17631] *******
[node01:17635] PMIX ERROR: PMIX_ERR_PACK_MISMATCH in file ../../src/client/pmix_client.c at line 278
[pspmix_0x0000000044df[0:17631]:0]: PMIx_Init failed: PMIX_ERROR
PSIlogger: Child with rank 0 exited with status 1.
```

> 2. Can you please walk me through your VM setup? I'd like to try and locally reproduce this so I can more efficiently track it down

I will try to, hope to make it tomorrow since next week, I'll be on vacation.

**Comment by rhc54** [comment]
> [node01:17635] PMIX ERROR: PMIX_ERR_PACK_MISMATCH in file ../../src/client/pmix_client.c at line 278

There's the problem!! Got a mismatch in what was packed vs what is being unpacked, so the bytes are all out of order. Now need to figure out what type was packed - and why we aren't seeing it elsewhere (you're the only one reporting a failure).

You can try adding the following patch:
```diff
diff --git a/src/mca/bfrops/base/bfrop_base_unpack.c b/src/mca/bfrops/base/bfrop_base_unpack.c
index b03fb923..af972486 100644
--- a/src/mca/bfrops/base/bfrop_base_unpack.c
+++ b/src/mca/bfrops/base/bfrop_base_unpack.c
@@ -54,7 +54,7 @@ static pmix_status_t pmix_bfrops_base_unpack_buffer(pmix_pointer_array_t *regtyp
         }
         /* if the data types don't match, then return an error */
         if (type != local_type) {
-            pmix_output_verbose(20, pmix_bfrops_base_framework.framework_output,
+           pmix_output(0,
                                 "PMIX bfrop:unpack: got type %s when expecting type %s",
                                 PMIx_Data_type_string(local_type),
                                 PMIx_Data_type_string(type));
```

**Comment by EloWork** [comment]
I just updates my test system and I was hitting this issue as well with pmix 5.3/5.4/5.6 while it works with pmix 4.2.9. I use VMs too. Maybe no one else is up to date?

**Comment by rhc54** [comment]
@EloWork Sorry for the very long delayed response. Can you please provide me with info on the configuration of your VM? Might allow me to reproduce this locally.
